src/node/cose_common.h

```diff
@@ -67,57 +72,168 @@ namespace ccf::cose
-    QCBORDecode_EnterMap(&ctx, NULL);
+    return {leaf_digest.h.begin(), leaf_digest.h.end()};
+  }
 
+  static void decode_receipt_top_level_phdr(
+    QCBORDecodeContext& ctx, CcfCoseReceiptPhdr& phdr)
+  {
     enum
     {
-      CWT_CLAIMS_INDEX,
+      ALG_INDEX,
+      KID_INDEX,
+      CWT_INDEX,
+      VDS_INDEX,
       END_INDEX,
     };
+
     QCBORItem header_items[END_INDEX + 1];
 
-    header_items[CWT_CLAIMS_INDEX].label.int64 = crypto::COSE_PHEADER_KEY_CWT;
-    header_items[CWT_CLAIMS_INDEX].uLabelType = QCBOR_TYPE_INT64;
-    header_items[CWT_CLAIMS_INDEX].uDataType = QCBOR_TYPE_MAP;
+    header_items[ALG_INDEX].label.int64 = ccf::crypto::COSE_PHEADER_KEY_ALG;
+    header_items[ALG_INDEX].uLabelType = QCBOR_TYPE_INT64;
+    header_items[ALG_INDEX].uDataType = QCBOR_TYPE_INT64;
+
+    header_items[KID_INDEX].label.int64 = ccf::crypto::COSE_PHEADER_KEY_ID;
+    header_items[KID_INDEX].uLabelType = QCBOR_TYPE_INT64;
+    header_items[KID_INDEX].uDataType = QCBOR_TYPE_BYTE_STRING;
+
+    header_items[CWT_INDEX].label.int64 = ccf::crypto::COSE_PHEADER_KEY_CWT;
+    header_items[CWT_INDEX].uLabelType = QCBOR_TYPE_INT64;
+    header_items[CWT_INDEX].uDataType = QCBOR_TYPE_MAP;
+
+    header_items[VDS_INDEX].label.int64 = ccf::crypto::COSE_PHEADER_KEY_VDS;
+    header_items[VDS_INDEX].uLabelType = QCBOR_TYPE_INT64;
+    header_items[VDS_INDEX].uDataType = QCBOR_TYPE_INT64;
 
     header_items[END_INDEX].uLabelType = QCBOR_TYPE_NONE;
 
     QCBORDecode_GetItemsInMap(&ctx, header_items);
 
-    qcbor_result = QCBORDecode_GetError(&ctx);
+    auto qcbor_result = QCBORDecode_GetError(&ctx);
     if (qcbor_result != QCBOR_SUCCESS)
     {
-      throw COSEDecodeError(
-        fmt::format("Failed to decode protected header: {}", qcbor_result));
+      throw ccf::cose::COSEDecodeError(fmt::format(
+        "Failed to decode protected header: {}",
+        qcbor_err_to_str(qcbor_result)));
     }
 
-    if (header_items[CWT_CLAIMS_INDEX].uDataType == QCBOR_TYPE_NONE)
+    if (header_items[ALG_INDEX].uDataType == QCBOR_TYPE_NONE)
     {
-      throw COSEDecodeError("Missing CWT claims in COSE_Sign1");
+      throw ccf::cose::COSEDecodeError(
+        "COSE receipt missing 'alg' in protected header");
     }
+    phdr.alg = header_items[ALG_INDEX].val.int64;
 
+    if (header_items[KID_INDEX].uDataType == QCBOR_TYPE_NONE)
+    {
+      throw ccf::cose::COSEDecodeError(
+        "COSE receipt missing 'kid' in protected header");
+    }
+    phdr.kid =
+      ccf::cose::qcbor_buf_to_byte_vector(header_items[KID_INDEX].val.string);
+
+    if (header_items[VDS_INDEX].uDataType == QCBOR_TYPE_NONE)
+    {
+      throw ccf::cose::COSEDecodeError(
+        "COSE receipt missing 'vds' in protected header");
+    }
+    phdr.vds = header_items[VDS_INDEX].val.int64;
+
+    if (phdr.vds != ccf::crypto::COSE_PHEADER_VDS_MERKLE_TREE)
+    {
+      throw ccf::cose::COSEDecodeError(fmt::format(
+        "Expected VDS={} (Merkle tree), got {}",
```

Comment:
```suggestion
        "Expected VDS={} (CCF_LEDGER_SHA256 Merkle tree), got {}",
```

there are several variants of Merkle Tree that can be used in COSE Receipts, in fact all profiles are Merkle Trees.